nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -1097,16 +1089,95 @@ impl<'a> BlueprintBuilder<'a> {
         }
 
-        // Remove any disks that appear in the blueprint, but not the database
-        for disk_id in blueprint_disk_ids {
-            if !database_disk_ids.contains(&disk_id) {
+        let final_counts = editor.edit_counts();
+        Ok(final_counts.difference_since(initial_counts))
+    }
+
+    /// Set the state of the disk to `Decommissioned` for any expunged disks if:
+    ///
+    /// (1) the sled where the disk resides has been expunged.
+    ///
+    ///   or
+    ///
+    /// (2) the sled-agent where the disk resides has seen the
+    /// request to expunge the disk as reflected in inventory. We use
+    /// `BlueprintPhysicalDisksConfig::generation` from the parent blueprint and
+    /// `SledAgent::omicron_physical_disks_generation` from inventory to see how
+    /// up to date the inventory is with respect to the parent_blueprint.
+    ///
+    ///
+    /// Caveats:
+    ///
+    ///   When the disk resides on an expunged sled, we can immediately
+    ///   decommission it, even if it has had its disposition set to expunged in
+    ///   *this* planner pass.
+    ///
+    ///   However, if the sled is still in service, we must wait for any
+    ///   disks expunged in this planner pass to be seen by the sled-agent and
+    ///   reflected in inventory before we can decommission them. Since any
+    ///   modifications made in this planner pass have not yet been sent to the
+    ///   sled-agents, we only check for expunged sleds in the parent_blueprint.
+    ///
+    ///
+    /// Called by the planner in the `do_plan_decommission()` stage of planning
+    pub fn sled_decommision_disks(
+        &mut self,
+        sled_id: SledUuid,
+        sled_details: &SledDetails,
+    ) -> Result<SledEditCounts, Error> {
+        let editor = self.sled_editors.get_mut(&sled_id).ok_or_else(|| {
+            Error::Planner(anyhow!(
+                "tried to decommission disks for unknown sled {sled_id}"
+            ))
+        })?;
+
+        let initial_counts = editor.edit_counts();
+
+        // This sled is expunged. Decommission all expunged disks. Note that
+        // because the decommision step comes after the expunge step, all disks
+        // should be marked expunged in the editor currently and can therefore
+        // be set to decommissioned.
+        //
+        // We shouldn't have added a disk during this planner pass, because
+        // an expunged sled is one that has been removed from the rack either
+        // physically or via trust-quorum.
+        if sled_details.policy == SledPolicy::Expunged {
+            for (_, disk) in sled_details.resources.all_disks(DiskFilter::All) {
                 editor
-                    .expunge_disk(&disk_id)
+                    .decommission_disk(&disk.disk_id)
                     .map_err(|err| Error::SledEditError { sled_id, err })?;
             }
+        } else if let Some(parent_bp_config) =
+            self.parent_blueprint.blueprint_disks.get(&sled_id)
```

Comment:
Do we still need to look at the parent blueprint now that we have `as_of_generation`? I think it's okay to look at our current disks (since if we just expunged them, the `as_of_generation` is guaranteed to be higher than anything that could be in inventory).

Reply:
Great catch. Fixed in 55a65f8d01ed8e54a7ade5b84eeae83a937956dd.